codegen/src/api/mod.rs

```diff
@@ -0,0 +1,360 @@
+    punctuated::Punctuated,
+};
+
+pub fn generate_runtime_api<P>(
+    item_mod: syn::ItemMod,
+    path: P,
+    generated_type_derives: Option<Punctuated<syn::Path, syn::Token![,]>>,
+) -> TokenStream2
+where
+    P: AsRef<path::Path>,
+{
+    let mut file = fs::File::open(&path).unwrap_or_else(|e| {
+        abort_call_site!("Failed to open {}: {}", path.as_ref().to_string_lossy(), e)
+    });
+
+    let mut bytes = Vec::new();
+    file.read_to_end(&mut bytes)
+        .unwrap_or_else(|e| abort_call_site!("Failed to read metadata file: {}", e));
+
+    let metadata = frame_metadata::RuntimeMetadataPrefixed::decode(&mut &bytes[..])
+        .unwrap_or_else(|e| abort_call_site!("Failed to decode metadata: {}", e));
+
+    let mut derives = GeneratedTypeDerives::default();
+    if let Some(user_derives) = generated_type_derives {
+        derives.append(user_derives.iter().cloned())
+    }
+
+    let generator = RuntimeGenerator::new(metadata);
+    generator.generate_runtime(item_mod, derives)
+}
+
+pub struct RuntimeGenerator {
+    metadata: RuntimeMetadataV14,
+}
+
+impl RuntimeGenerator {
+    pub fn new(metadata: RuntimeMetadataPrefixed) -> Self {
+        match metadata.1 {
+            RuntimeMetadata::V14(v14) => Self { metadata: v14 },
+            _ => panic!("Unsupported metadata version {:?}", metadata.1),
+        }
+    }
+
+    pub fn generate_runtime(
+        &self,
+        item_mod: syn::ItemMod,
+        derives: GeneratedTypeDerives,
+    ) -> TokenStream2 {
+        let item_mod_ir = ir::ItemMod::from(item_mod);
+
+        // some hardcoded default type substitutes, can be overridden by user
+        let mut type_substitutes = [
+            (
+                "bitvec::order::Lsb0",
+                parse_quote!(::subxt::bitvec::order::Lsb0),
+            ),
+            (
+                "bitvec::order::Msb0",
+                parse_quote!(::subxt::bitvec::order::Msb0),
+            ),
+            (
+                "sp_core::crypto::AccountId32",
+                parse_quote!(::subxt::sp_core::crypto::AccountId32),
+            ),
+            (
+                "primitive_types::H256",
+                parse_quote!(::subxt::sp_core::H256),
+            ),
+            (
+                "sp_runtime::multiaddress::MultiAddress",
+                parse_quote!(::subxt::sp_runtime::MultiAddress),
+            ),
+            // todo: [AJ] remove the requirement for these by implementing Compact handling properly
+            (
+                "sp_arithmetic::per_things::Perbill",
+                parse_quote!(::subxt::sp_arithmetic::per_things::Perbill),
+            ),
+            (
+                "sp_arithmetic::per_things::Perquintill",
+                parse_quote!(::subxt::sp_arithmetic::per_things::Perquintill),
+            ),
+            (
+                "frame_support::traits::misc::WrapperKeepOpaque",
+                parse_quote!(::subxt::WrapperKeepOpaque),
+            ),
+        ]
+        .iter()
```

Comment:
I think we can use `into_iter()` and avoid the cloning now? :)

Reply:
(and perhaps we don't need the `.to_string()` below either; leave as str?)

(just nits, really)

Reply:
#309